src/vote.rs

```diff
@@ -0,0 +1,135 @@
+use crate::{Error, Height, Keychain, Round, Step};
+
+use fuel_crypto::{Hasher, PublicKey, SecretKey, Signature};
+use fuel_types::Bytes32;
+
+/// A vote from a validator.
+///
+/// These votes are consumed to produce state change in the reactor.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+pub struct Vote {
```

Comment:
If I'm going to send this over the wire I would need this struct and it's fields to implement serde's Serialize & Deserialize.

But I'm thinking sending `validator: PublicKey` over the network would be redundant since we already know which peer we're talking to. Having said that I would need to store PublicKey of the connected peer, cuz currently in p2p we only a multihash of the PublicKey aka PeerId.

Reply:
I think the peering/libp2p keys should be a separate and independent key from the consensus keys.

Reply:
makes sense, either way p2p-service needs to be either aware of those keys or we send it with each `BroadcastConensus(Vote)` message.

We can share the keys on initial connection / hello request

Reply:
hmm since it's a gossipsub protocol, maybe it would be better to have the key sent over every time, we might not be aware of the PublicKey of the peer we received message from

Reply:
This information is already available via the signature field. We can recover the consensus public key from the signature field and then compare that with the mapping in the staking contract to determine which validator it is.

Reply:
should we then remove it from Vote struct altogether ?

Reply:
I think so, it just opens up a vector for griefing imo

Reply:
@vlopes11 let us know what you think and if we're going to send Vote as is, I would need it's field to be de/serializable

Reply:
> We can recover the consensus public key from the signature field

Note that this is a detail of the specific signature scheme and curve that's used, but may not hold for other schemes or curves.

Reply:
Unless we plan to support some way to obscure the identity of validators, then I think we should build our protocol around only supporting signature schemes that do support public key recovery. Otherwise there's no way to prove that a block proposal came from the correct sender, or guarantee that you're slashing the correct validator (if applicable).

Reply:
Oh I meant that e.g. EdDSA doesn't allow you to _recover_ the pubkey but does allow you to verify a signature _given_ the pubkey. EdDSA is used in many many chains. We shouldn't assume the signature scheme allows recovery of the pubkey.

Reply:
Ahh good points, I was thinking in terms of ecdsa and basic schnorr identification scheme. EdDSA and hardened signatures are definitely something that should be supportable in this library, so I take back what I said earlier.

Reply:
IIUC there is no action to be taken from this thread. Should we leave an open issue to make the signer generic and accept different curves?

Reply:
I think opening an issue to make signature generic is a good idea.

Reply:
I think opening an issue to make signature generic is a good idea.